README.md
# fuel

Client for ExactTarget's Fuel REST APIs

[![Build Status](https://travis-ci.org/ExactTarget/node-fuel.png?branch=master)](https://travis-ci.org/ExactTarget/node-fuel)

## Getting Started
Install the module with: `npm install fuel`

### Call a Fuel REST API

```javascript
// Load and configure the module

var fuel = require('fuel').configure({
	authUrl: 'https://auth.exacttargetapis.com/v1/requestToken',
	clientId: 'xxxxxxxxxxxxxxxxxxxxxxxx',
	clientSecret: 'yyyyyyyyyyyyyyyyyyyyyyyy'
});

// OR, if you have a refresh token

var fuel = require('fuel').configure({
	authUrl: 'https://auth.exacttargetapis.com/v1/requestToken',
	clientId: 'xxxxxxxxxxxxxxxxxxxxxxxx',
	clientSecret: 'yyyyyyyyyyyyyyyyyyyyyyyy',
	refreshToken: 'zzzzzzzzzzzzzzzzzzzzzzzz',
	accessType: 'offline'
});

// The fuel module will manage your access token behind the
// scenes, renewing it when necessary, maintaining state
// using the refresh token if present

// Call the API (this example displays your user context)

fuel({
	url: 'https://www.exacttargetapis.com/platform/v1/tokenContext'
}, function (error, request, body) {
	console.log(body);
});
```
#### Syntax

The general format is as follows:

`fuel(options, callback);`

The `options` and `callback` parameters are compatible with the `request` module.  For details, see the documentation:

https://github.com/mikeal/request#requestoptions-callback

### Just manage a Fuel OAuth token

```javascript
// Load and configure the module

var token = require('fuel').token.configure({
	authUrl: 'https://auth.exacttargetapis.com/v1/requestToken',
	clientId: 'xxxxxxxxxxxxxxxxxxxxxxxx',
	clientSecret: 'yyyyyyyyyyyyyyyyyyyyyyyy'
});

// OR, if you have a refresh token

var token = require('fuel').token.configure({
	authUrl: 'https://auth.exacttargetapis.com/v1/requestToken',
	clientId: 'xxxxxxxxxxxxxxxxxxxxxxxx',
	clientSecret: 'yyyyyyyyyyyyyyyyyyyyyyyy',
	refreshToken: 'zzzzzzzzzzzzzzzzzzzzzzzz',
	accessType: 'offline'
});

// Get a token (this example displays the token data)

token(function (error, response, tokenData) {
	console.log(tokenData);
});

// Repeated calls to the token function will return a cached
// token. This module will manage your access token behind the
// scenes, renewing it when necessary, maintaining state
// using the refresh token if present
```

#### Syntax

The general format is as follows:

`token(callback);`

The `callback` parameter is compatible with the `request` module.  For details, see the documentation:

https://github.com/mikeal/request#requestoptions-callback

## Contributing
Before writing code, we suggest you [search for issues](https://github.com/ExactTarget/node-mashery/issues?state=open)
or [create a new one](https://github.com/ExactTarget/node-mashery/issues/new) to confirm where your contribution fits into
our roadmap.

In lieu of a formal style guide, take care to maintain the existing coding style. Add unit tests for any new or changed functionality.
Lint and test your code using [grunt](https://github.com/cowboy/grunt).

## Acknowledgements

We are grateful to the following maintainers, contributors, and sponsors of the technologies used by this module.

* [Node.js](http://nodejs.org)

* [Request](https://github.com/mikeal/request) (De facto HTTP request module for Node)

* [grunt](https://github.com/cowboy/grunt) (Build tool for JavaScript projects)

##Authors

**Adam Alexander**

+ http://twitter.com/adamalex
+ http://github.com/adamalex

**Benjamin Dean**

+ https://twitter.com/bdeanet
+ https://github.com/creatovisguru

**Douglas Wilson**

+ https://twitter.com/blipsofadoug
+ https://github.com/dougwilson

## Copyright and license

Copyright (c) 2013 ExactTarget

Licensed under the MIT License (the "License");
you may not use this work except in compliance with the License.
You may obtain a copy of the License in the COPYING file.

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


History.md
fuel History
============

v0.6.0 2015-08-27
-----------------

This release removes credentials length validations, upgrades `request`,
and drops support for Node.js 0.8.

  * Remove length check from `clientId` and `clientSecret` options
  * Remove support for Node.js 0.8
  * `request` module updated to v2.61.0 from v2.45.0.

v0.5.0 2015-08-21
-----------------

This release is a maintainance release to upgrade dependencies.

  * `extend` module updated to v3.0.0 from v1.2.1.
  * `request` module updated to v2.45.0 from v2.27.0.

v0.4.1 2014-01-21
-----------------

This release fixes an issue where an error response from the token request
will cache the invalid state forever.

  * Forward error from the request to the callback.
  * Only cache token requests when the response body contains an `expiresIn`
    property (i.e. do not cache error responses).

v0.4.0 2013-12-03
-----------------

This release adds better token handling when using automatic token
management.

  * `request` module updated to v2.27.0 from v0.12.0.
  * Return error when fetching token automatically for request fails.
  * Attempt to fetch new token after a request fails with invalid token
    response.

v0.3.1 2013-10-17
-----------------

This release fixes a regression in v0.3.0 where tokens would never expire
and so after token expiration time was reached, all requests would result
in `401 Unauthorized`.

  *  Fix omission preventing memorized tokens from expiring.

v0.3.0 2013-10-17
-----------------

This release removes the token expiration timer. The main benefit is that
the expiration will occur even if the event loop is being starved or the
clock drifts.

  * Replace timer in token response memorization with on-demand checks.
  * Use monotonic clock for memorization expiration rather than wall clock.

v0.2.3 2013-10-08
-----------------

This release contains a fix for a regression in `v0.2.2` that broke
compatibility with node < 0.10.

  * Fix compatibility with Node.js below v0.10.

v0.2.2 2013-06-11
-----------------

  * Prevent token expiration timer from artificially keeping the event
    loop open.

v0.2.1 2013-03-20
-----------------

  * Copy stored options to avoid collision.

v0.2.0 2013-02-02
-----------------

  * Improve validation of options.
  * Support arrays within request body.
  * Cache OAuth errors for 10 seconds.

v0.1.0 2013-01-26
-----------------

This is the first, initial implementation.


test/fuel_test.js
var fuel = require('../lib/fuel.js');

/*
 ======== A Handy Little Nodeunit Reference ========
 https://github.com/caolan/nodeunit

 Test methods:
 test.expect(numAssertions)
 test.done()
 Test assertions:
 test.ok(value, [message])
 test.equal(actual, expected, [message])
 test.notEqual(actual, expected, [message])
 test.deepEqual(actual, expected, [message])
 test.notDeepEqual(actual, expected, [message])
 test.strictEqual(actual, expected, [message])
 test.notStrictEqual(actual, expected, [message])
 test.throws(block, [error], [message])
 test.doesNotThrow(block, [error], [message])
 test.ifError(value)
 */

exports['module basics'] = {
	'initializes': function (test) {
		test.expect(4);
		test.equal(typeof fuel, 'function', 'fuel should be a function');
		test.equal(typeof fuel.configure, 'function', 'fuel.configure should be a function');
		test.equal(typeof fuel.token, 'function', 'fuel.token should be a function');
		test.equal(typeof fuel.token.configure, 'function', 'fuel.token should be a function');
		test.done();
	},

	'validates options': function (test) {
		test.expect(2);
		fuel({}, function (error) {
			test.ok(error instanceof Error, 'should call callback with error object');
			test.equal(typeof error.message, 'string', 'error should contain a string message');
			test.done();
		});
	}
};

exports['module validation'] = {
	setUp: function (done) {
		stubCallback.callbackCount = 0;
		stubCallback.error = null;
		done();
	},

	tearDown: function (done) {
		stubCallback.callbackCount = 0;
		stubCallback.error = null;
		done();
	},

	'validates null options': function (test) {
		var valid = fuel._validateOptions(null, stubCallback);

		test.ok(stubCallback.error instanceof Error, 'should call callback with error object');
		test.strictEqual(stubCallback.error.message, 'options object is required', 'should return expected message');
		test.strictEqual(stubCallback.callbackCount, 1, 'should call the callback once');
		test.strictEqual(valid, false, 'should indicate validation failure');
		test.done();
	},

	'validates empty options': function (test) {
		var valid = fuel._validateOptions({}, stubCallback);

		test.ok(stubCallback.error instanceof Error, 'should call callback with error object');
		test.strictEqual(stubCallback.error.message, 'authUrl is required', 'should return expected message');
		test.strictEqual(stubCallback.callbackCount, 1, 'should call the callback once');
		test.strictEqual(valid, false, 'should indicate validation failure');
		test.done();
	},

	'validates missing clientId': function (test) {
		var valid = fuel._validateOptions({ authUrl: 'a' }, stubCallback);

		test.ok(stubCallback.error instanceof Error, 'should call callback with error object');
		test.strictEqual(stubCallback.error.message, 'clientId is missing or invalid', 'should return expected message');
		test.strictEqual(stubCallback.callbackCount, 1, 'should call the callback once');
		test.strictEqual(valid, false, 'should indicate validation failure');
		test.done();
	},

	'validates invalid clientId': function (test) {
		var valid = fuel._validateOptions({ authUrl: 'a', clientId: [] }, stubCallback);

		test.ok(stubCallback.error instanceof Error, 'should call callback with error object');
		test.strictEqual(stubCallback.error.message, 'clientId is missing or invalid', 'should return expected message');
		test.strictEqual(stubCallback.callbackCount, 1, 'should call the callback once');
		test.strictEqual(valid, false, 'should indicate validation failure');
		test.done();
	},

	'validates missing clientSecret': function (test) {
		var valid = fuel._validateOptions({ authUrl: 'a', clientId: 'iiiiiiiiiiiiiiiiiiiiiiii' }, stubCallback);

		test.ok(stubCallback.error instanceof Error, 'should call callback with error object');
		test.strictEqual(stubCallback.error.message, 'clientSecret is missing or invalid', 'should return expected message');
		test.strictEqual(stubCallback.callbackCount, 1, 'should call the callback once');
		test.strictEqual(valid, false, 'should indicate validation failure');
		test.done();
	},

	'validates invalid clientSecret': function (test) {
		var valid = fuel._validateOptions({ authUrl: 'a', clientId: 'iiiiiiiiiiiiiiiiiiiiiiii', clientSecret: [] }, stubCallback);

		test.ok(stubCallback.error instanceof Error, 'should call callback with error object');
		test.strictEqual(stubCallback.error.message, 'clientSecret is missing or invalid', 'should return expected message');
		test.strictEqual(stubCallback.callbackCount, 1, 'should call the callback once');
		test.strictEqual(valid, false, 'should indicate validation failure');
		test.done();
	},

	'validates valid options': function (test) {
		var valid = fuel._validateOptions({
			authUrl: 'a',
			clientId: 'iiiiiiiiiiiiiiiiiiiiiiii',
			clientSecret: 'ssssssssssssssssssssssss'
		}, stubCallback);

		test.strictEqual(stubCallback.callbackCount, 0, 'should not call the callback');
		test.strictEqual(valid, true, 'should indicate validation success');
		test.done();
	}
};

exports['fuel requests and config'] = {
	setUp: function (done) {
		this._performRequest = fuel._performRequest;
		fuel._performRequest = stubEchoRequest;
		this._performTokenRequest = fuel._performTokenRequest;
		fuel._performTokenRequest = stubTokenRequest;
		done();
	},

	tearDown: function (done) {
		fuel._performRequest = this._performRequest;
		fuel._performTokenRequest = this._performTokenRequest;
		done();
	},

	'handles a basic request': function (test) {
		test.expect(2);

		fuel({
			url: 'apiurl',
			method: 'PATCH',
			authUrl: 'auth',
			clientId: 'yyyyyyyyyyyyyyyyyyyyyyyy',
			clientSecret: 'zzzzzzzzzzzzzzzzzzzzzzzz',
			body: 'zzzzz',
			accessType: 'aaaa',
			refreshToken: 'rrrr',
			scope: 'ssss'
		}, function (error, response, body) {
			test.ifError(error);

			test.deepEqual(body, {
				url: 'apiurl',
				method: 'PATCH',
				json: true,
				headers: {
					authorization: 'Bearer stubtoken'
				},
				body: 'zzzzz'
			}, 'should construct the appropriate request');

			test.done();
		});
	},

	'handles invalid token response': function (test) {
		test.expect(2);

		fuel({
			url: 'apiurl',
			method: 'PATCH',
			authUrl: 'authbad',
			clientId: 'yyyyyyyyyyyyyyyyyyyyyyyy',
			clientSecret: 'zzzzzzzzzzzzzzzzzzzzzzzz',
			body: 'zzzzz',
			accessType: 'aaaa',
			refreshToken: 'rrrr',
			scope: 'ssss'
		}, function (error, response, body) {
			test.ifError(error);

			test.deepEqual(body, {
				url: 'apiurl',
				method: 'PATCH',
				json: true,
				headers: {
					authorization: 'Bearer stubtoken'
				},
				body: 'zzzzz'
			}, 'should construct the appropriate request');

			test.done();
		});
	},

	'manages separate configurations': function (test) {
		test.expect(4);

		var fuela = fuel.configure({
			authUrl: 'auth',
			clientId: 'yyyyyyyyyyyyyyyyyyyyyyyy',
			clientSecret: 'zzzzzzzzzzzzzzzzzzzzzzzz',
			accessType: 'aaaa',
			refreshToken: 'rrrr',
			scope: 'ssss'
		});

		var fuelx = fuel.configure({
			authUrl: 'auth2',
			clientId: 'yyyyyyyyyyyyyyyyyyyyyyyy',
			clientSecret: 'zzzzzzzzzzzzzzzzzzzzzzzz',
			accessType: 'aaaa',
			refreshToken: 'rrrr',
			scope: 'ssss'
		});

		fuela({ url: 'apiurla', method: 'PATCH', body: 'zzzzz' }, function (error, response, body) {
			test.ifError(error);

			test.deepEqual(body, {
				url: 'apiurla',
				method: 'PATCH',
				json: true,
				headers: {
					authorization: 'Bearer stubtoken'
				},
				body: 'zzzzz'
			}, 'should construct the appropriate fuela request');
		});

		fuelx({ url: 'apiurlx', method: 'PATCH', body: 'zzzzz' }, function (error, response, body) {
			test.ifError(error);

			test.deepEqual(body, {
				url: 'apiurlx',
				method: 'PATCH',
				json: true,
				headers: {
					authorization: 'Bearer stubtoken2'
				},
				body: 'zzzzz'
			}, 'should construct the appropriate fuelx request');

			test.done();
		});
	},

	'manages multiple requests without colliding': function (test) {
		test.expect(4);

		var fuela = fuel.configure({
			authUrl: 'auth',
			clientId: 'yyyyyyyyyyyyyyyyyyyyyyyy',
			clientSecret: 'zzzzzzzzzzzzzzzzzzzzzzzz',
			accessType: 'aaaa',
			refreshToken: 'rrrr',
			scope: 'ssss'
		});

		fuela({ url: 'apiurla', method: 'PATCH', body: { one: true } }, function (error, response, body) {
			test.ifError(error);

			test.deepEqual(body, {
				url: 'apiurla',
				method: 'PATCH',
				json: true,
				headers: {
					authorization: 'Bearer stubtoken'
				},
				body: { one: true }
			}, 'should construct the appropriate fuela request');
		});

		fuela({ url: 'apiurla', method: 'PATCH', body: { two: true } }, function (error, response, body) {
			test.ifError(error);

			test.deepEqual(body, {
				url: 'apiurla',
				method: 'PATCH',
				json: true,
				headers: {
					authorization: 'Bearer stubtoken'
				},
				body: { two: true }
			}, 'should construct the appropriate fuela request');
		});

		test.done();
	},

	'handles token request 500 error': function (test) {
		test.expect(2);

		fuel({
			url: 'apiurl',
			method: 'PATCH',
			authUrl: 'auth500',
			clientId: 'yyyyyyyyyyyyyyyyyyyyyyyy',
			clientSecret: 'zzzzzzzzzzzzzzzzzzzzzzzz',
			body: 'zzzzz',
			accessType: 'aaaa',
			refreshToken: 'rrrr',
			scope: 'ssss'
		}, function (error) {
			test.ok(error, 'should get error');
			test.equal(error.message, 'Error requesting token: Internal Server Error', 'should be expected error');
			test.done();
		});
	},

	'handles token request non-200': function (test) {
		test.expect(2);

		fuel({
			url: 'apiurl',
			method: 'PATCH',
			authUrl: 'auth503',
			clientId: 'yyyyyyyyyyyyyyyyyyyyyyyy',
			clientSecret: 'zzzzzzzzzzzzzzzzzzzzzzzz',
			body: 'zzzzz',
			accessType: 'aaaa',
			refreshToken: 'rrrr',
			scope: 'ssss'
		}, function (error) {
			test.ok(error, 'should get error');
			test.equal(error.message, 'Error requesting token: 503 Service Unavailable', 'should be expected error');
			test.done();
		});
	},

	'handles token request bad response': function (test) {
		test.expect(2);

		fuel({
			url: 'apiurl',
			method: 'PATCH',
			authUrl: 'authempty',
			clientId: 'yyyyyyyyyyyyyyyyyyyyyyyy',
			clientSecret: 'zzzzzzzzzzzzzzzzzzzzzzzz',
			body: 'zzzzz',
			accessType: 'aaaa',
			refreshToken: 'rrrr',
			scope: 'ssss'
		}, function (error) {
			test.ok(error, 'should get error');
			test.equal(error.message, 'Error requesting token: Token Missing', 'should be expected error');
			test.done();
		});
	}
};

exports['fuel token and config'] = {
	setUp: function (done) {
		this._performRequest = fuel._performRequest;
		fuel._performRequest = stubEchoRequest;
		this._performTokenRequest = fuel._performTokenRequest;
		fuel._performTokenRequest = stubEchoTokenRequest;
		done();
	},

	tearDown: function (done) {
		fuel._performRequest = this._performRequest;
		fuel._performTokenRequest = this._performTokenRequest;
		done();
	},

	'handles a token request': function (test) {
		test.expect(2);

		fuel.token({
			authUrl: 'auth3',
			clientId: 'yyyyyyyyyyyyyyyyyyyyyyyy',
			clientSecret: 'zzzzzzzzzzzzzzzzzzzzzzzz',
			accessType: 'aaaa',
			refreshToken: 'rrrr',
			scope: 'ssss'
		}, function (error, response, body) {
			test.ifError(error);

			test.deepEqual(body, {
				url: 'auth3',
				method: 'POST',
				json: true,
				body: {
					clientId: 'yyyyyyyyyyyyyyyyyyyyyyyy',
					clientSecret: 'zzzzzzzzzzzzzzzzzzzzzzzz',
					accessType: 'aaaa',
					refreshToken: 'rrrr',
					scope: 'ssss'
				}
			}, 'should construct the appropriate request');

			test.done();
		});
	},

	'manages separate configurations': function (test) {
		test.expect(4);

		var tokena = fuel.token.configure({
			authUrl: 'auth3',
			clientId: 'aaaaaaaaaaaaaaaaaaaaaaaa',
			clientSecret: 'bbbbbbbbbbbbbbbbbbbbbbbb',
			accessType: 'aaaa',
			refreshToken: 'rrrr',
			scope: 'ssss'
		});

		var tokenx = fuel.token.configure({
			authUrl: 'auth4',
			clientId: 'xxxxxxxxxxxxxxxxxxxxxxxx',
			clientSecret: 'yyyyyyyyyyyyyyyyyyyyyyyy',
			accessType: 'aaaa',
			refreshToken: 'rrrr',
			scope: 'ssss'
		});

		tokena(function (error, response, body) {
			test.ifError(error);

			test.deepEqual(body, {
				url: 'auth3',
				method: 'POST',
				json: true,
				body: {
					clientId: 'aaaaaaaaaaaaaaaaaaaaaaaa',
					clientSecret: 'bbbbbbbbbbbbbbbbbbbbbbbb',
					accessType: 'aaaa',
					refreshToken: 'rrrr',
					scope: 'ssss'
				}
			}, 'should construct the appropriate tokena request');
		});

		tokenx(function (error, response, body) {
			test.ifError(error);

			test.deepEqual(body, {
				url: 'auth4',
				method: 'POST',
				json: true,
				body: {
					clientId: 'xxxxxxxxxxxxxxxxxxxxxxxx',
					clientSecret: 'yyyyyyyyyyyyyyyyyyyyyyyy',
					accessType: 'aaaa',
					refreshToken: 'rrrr',
					scope: 'ssss'
				}
			}, 'should construct the appropriate tokenx request');

			test.done();
		});
	}
};


// Stub logic

function stubCallback(error) {
	stubCallback.callbackCount += 1;
	stubCallback.error = error;
}

function stubEchoRequest(options, callback) {
	var body = fuel._removeAuthOptions(options);
	var res = {"statusCode":200};

	if (options.headers.authorization === 'Bearer badtoken') {
		body = {};
		res.statusCode = 401;
		res.headers = {"www-authenticate":"Bearer realm=\"test\", error=\"invalid_token\""};
		options.authUrl = 'auth'; // Set back to good authentication
	}

	callback(null, res, body);
}

function stubEchoTokenRequest(options, callback) {
	callback(null, {}, options);
}

function stubTokenRequest(options, callback) {
	var body = {};
	var res = {"statusCode":200};

	switch (options.url) {
		case 'auth':
			body.accessToken = 'stubtoken';
			body.expiresIn = 1;
			break;
		case 'auth2':
			body.accessToken = 'stubtoken2';
			body.expiresIn = 1;
			break;
		case 'auth500':
			body.documentation = 'https://code.docs.exacttarget.com/rest/errors/500';
			body.errorcode = 0;
			body.message = 'Internal Server Error';
			res.statusCode = 500;
			break;
		case 'auth503':
			body = null;
			res.statusCode = 503;
			break;
		case 'authbad':
			body.accessToken = 'badtoken';
			body.expiresIn = 1;
			break;
		case 'authempty':
			break;
	}

	callback(null, res, body);
}


lib/fuel.js
/*
 * fuel
 * https://github.com/ExactTarget/node-fuel
 *
 * Copyright (c) 2013 ExactTarget
 * Licensed under the MIT license.
 */

var extend = require('extend');
var http = require('http');
var request = require('request');

var invalidTokenRegExp = /^Bearer\s.+?invalid_token/;

function fuel(options, callback) {
	var retries = 1;
	function makeRequest(options, callback) {
		fuel.token(options, function onTokenResponse(error, response, body) {
			if (error) return callback(error);
			if (response.statusCode !== 200) return callback(new Error('Error requesting token: ' + ((body && body.message) || (response.statusCode + ' ' + http.STATUS_CODES[response.statusCode]))));
			if (!body || !body.accessToken) return callback(new Error('Error requesting token: Token Missing'));
			options.json = true;
			options.headers = options.headers || {};
			options.headers.authorization = 'Bearer ' + body.accessToken;

			fuel._performRequest(options, function onResponse(error, res) {
				if (error) return callback(error);
				if (res.statusCode !== 401) return callback.apply(null, arguments);
				if ((retries -= 1) < 0) return callback.apply(null, arguments);
				if (!res.headers || !res.headers['www-authenticate']) return callback.apply(null, arguments);
				if (!invalidTokenRegExp.test(res.headers['www-authenticate'])) return callback.apply(null, arguments);
				if (fuel.token.memorized) {
					fuel.token.clear(fuel._authOptionsHash(createAuthOptions(options)));
				}
				makeRequest(options, callback);
			});
		});
	}
	makeRequest(options, callback);
}

fuel.token = function (options, callback) {

	if (!fuel._validateOptions(options, callback)) return;

	var authOptions = createAuthOptions(options);
	var authOptionsHash = fuel._authOptionsHash(authOptions);

	fuel._performTokenRequest(authOptions, function (error, response, body) {
		if (error) {
			callback(error);
		} else {
			if (body && body.refreshToken) {
				fuel._refreshTokenCache[authOptionsHash] = body.refreshToken;
			}
			callback(error, response, body);
		}
	});
};

Object.defineProperty(fuel.token, 'memorized', {
	'get': function () { return fuel._performTokenRequest.memorized; }
});

fuel.configure = createConfigurationWrapper(fuel);
fuel.token.configure = createConfigurationWrapper(fuel.token);

fuel._validateOptions = function (options, callback) {
	var message = null;

	// When validating, handle the negations first
	if (!options) {
		message = 'options object is required';
	} else if (!options.authUrl) {
		message = 'authUrl is required';
	} else if (!options.clientId || typeof options.clientId !== 'string') {
		message = 'clientId is missing or invalid';
	} else if (!options.clientSecret || typeof options.clientSecret !== 'string') {
		message = 'clientSecret is missing or invalid';
	}

	// Something is wrong, notifiy the user
	if (null !== message) {
		callback(new Error(message));
		return false;
	} else {
		// Should only be here if all is well
		return true;
	}
};

fuel._removeAuthOptions = function (options) {
	options = extend(true, {}, options);
	delete options.clientId;
	delete options.clientSecret;
	delete options.refreshToken;
	delete options.accessType;
	delete options.authUrl;
	delete options.scope;
	return options;
};

fuel._performRequest = function (options, callback) {
	return request(fuel._removeAuthOptions(options), callback);
};

fuel._performTokenRequest = memoize(fuel._performRequest, function (options) {
	// Cache response per unique set of options
	return fuel._authOptionsHash(options);
}, function (error, response, body) {
	// Cache tokens until 30 seconds before expiration
	return body && (body.expiresIn - 30) || 0;
});
fuel.token.clear = fuel._performTokenRequest.clear.bind(fuel._performTokenRequest);

fuel._authOptionsHash = function (options) {
	return options.url + options.body.clientId + options.body.scope;
};

fuel._refreshTokenCache = {};

module.exports = fuel;


// Utility

function createAuthOptions(options) {
	var authOptions = {
		'url': options.authUrl,
		'method': 'POST',
		'json': true,
		'body': {
			'clientId': options.clientId,
			'clientSecret': options.clientSecret
		}
	};

	if (options.accessType) authOptions.body.accessType = options.accessType;
	if (options.refreshToken) authOptions.body.refreshToken = options.refreshToken;
	if (options.scope) authOptions.body.scope = options.scope;

	var authOptionsHash = fuel._authOptionsHash(authOptions);

	if (fuel._refreshTokenCache[authOptionsHash]) {
		authOptions.body.refreshToken = fuel._refreshTokenCache[authOptionsHash];
	}

	return authOptions;
}

function createConfigurationWrapper(func) {
	return function (storedOptions) {
		return function (options, callback) {
			if (!callback) {
				callback = options;
				options = {};
			}
			return func(extend(true, {}, storedOptions, options), callback);
		};
	};
}

function memoize(fn, hasher, invalidator) {
	var exp = {};
	var memo = {};
	var queues = {};
	hasher = hasher || function (x) {
		return x;
	};
	invalidator = invalidator || function () {};
	var clear = function (key) {
		delete exp[key];
		delete memo[key];
	};
	var memoized = function () {
		var args = Array.prototype.slice.call(arguments);
		var callback = args.pop();
		var key = hasher.apply(null, args);
		reap();
		if (key in memo) {
			callback.apply(null, memo[key]);
		}
		else if (key in queues) {
			queues[key].push(callback);
		}
		else {
			queues[key] = [callback];
			fn.apply(null, args.concat([function () {
				var timeout = invalidator.apply(null, arguments);
				if (typeof timeout === 'number') {
					exp[key] = process.hrtime()[0] + timeout;
					memo[key] = arguments;
				}
				var q = queues[key];
				delete queues[key];
				for (var i = 0, l = q.length; i < l; i += 1) {
					q[i].apply(null, arguments);
				}
			}]));
		}
	};
	var reap = function () {
		var i;
		var key;
		var keys = Object.keys(memo);
		var length;
		var now = process.hrtime()[0];

		for (i = 0, length = keys.length; i < length; i += 1) {
			key = keys[i];
			if (key in exp && exp[key] <= now) {
				delete exp[key];
				delete memo[key];
			}
		}
	};
	memoized.clear = clear;
	memoized.memorized = true;
	memoized.unmemoized = fn;
	return memoized;
}
